### PR TITLE
Refactor proof generation functions

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This change introduces the `proof` package which helps to generate
inclusion and consistency proofs. The new `proof.Nodes` type replaces the
old `[]NodeFetch` and the `Rehash` function.

Key improvements:

- Uses `[]compact.NodeID` natively, no need to convert from `[]NodeFetch` up the stack.
- Stricter handling of ephemeral nodes: `proof.Nodes` allows only one whereas a
  `[]NodeFetch` could have arbitrary runs of rehashes.
- Concise naming: `proof.Inclusion` and `proof.Consistency` return `proof.Nodes`.
- Shorter code.

Addresses #3. Repeats https://github.com/google/trillian/pull/2576.

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
It might be easy to review commit by commit to follow some logic. There are a few mechanical copy-paste commits.

File: proof/proof.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"math/bits"
+
+	"github.com/transparency-dev/merkle/compact"
+)
+
+// Nodes contains information on how to construct a log Merkle tree proof.
```
Comment (AlCutter) [inline proof/proof.go:26]: Probably worth expanding the comment here to say something along the lines that the proof built from these nodes can have at most 1 ephemeral node?

File: proof/proof.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+type Nodes struct {
+	// IDs contains the IDs of non-ephemeral nodes sufficient to build the proof.
+	IDs []compact.NodeID
+	// begin is the beginning index (inclusive) into the IDs[begin:end] subslice
+	// of the nodes comprising the ephemeral node.
```
Comment (AlCutter) [inline proof/proof.go:31]: language nit: Maybe change "comprising" to something like "... nodes which will be used to recreate the ephemeral node"? (To me at least, "comprise" has the connotation that some things are directly present within a greater thing, whereas here we're deriving a new hash from a set of inputs which feels slightly different)

File: proof/proof.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	index := (size1 - 1) >> level
+	// If it does not cover the whole size1 tree, add this node to the proof.
+	if index != 0 {
+		proof = append(proof, compact.NewNodeID(level, index))
+	}
+
+	// Now append the path from this node to the root of size2.
+	p := nodes(index, level, size2)
+	p.IDs = append(proof, p.IDs...)
+	if len(proof) == 1 && p.begin < p.end {
```
Comment (AlCutter) [inline proof/proof.go:71]: Maybe just add a comment here to explain that this is adjusting for when we already have a node in the proof for the cases where size1 isn't a power of two (and the verifier needs to be able to re-create the root hash at size1)?

File: proof/proof.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+		p.begin++
+		p.end++
+	}
+	return p, nil
+}
+
+// nodes returns the node IDs necessary to prove that the (level, index) node
+// is included in the Merkle tree of the given size.
+func nodes(index uint64, level uint, size uint64) Nodes {
+	// [begin, end) is the leaves range covered by the (level, index) node.
```
Comment (AlCutter) [inline proof/proof.go:81]: leaves -> leaf

File: proof/proof.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+// Rehash computes the proof based on the slice of node hashes corresponding to
+// their IDs in the n.IDs field. The slices must be of the same length. The hc
+// parameter computes a node's hash based on hashes of its children.
+//
+// Warning: The passed-in slice of hashes can be modified in-place.
+func (n Nodes) Rehash(h [][]byte, hc func(left, right []byte) []byte) ([][]byte, error) {
+	if len(h) != len(n.IDs) {
+		return nil, errors.New("slice lengths mismatch")
```
Comment (AlCutter) [inline proof/proof.go:135]: Maybe be a bit more explicit in the error here to help debugging? (e.g. `fmt.Errorf("got %d hashes but expected %d", len(h), len(n.IDs)` or something along those lines?)

**Comment by AlCutter** [comment]
LGTM, @mhutchinson would you like to have a look over?

**Review by AlCutter — APPROVED** [review]

**Pull request merged**